src/audio/multiband_drc/multiband_drc.c

```diff
@@ -0,0 +1,605 @@
+
+	return 0;
+}
+
+static int multiband_drc_crossover_eq_init_coef(struct multiband_drc_comp_data *cd, int16_t nch)
+{
+	struct sof_eq_iir_biquad_df2t *crossover;
+	struct sof_eq_iir_biquad_df2t *emphasis;
+	struct sof_eq_iir_biquad_df2t *deemphasis;
+	struct sof_multiband_drc_config *config = cd->config;
+	struct multiband_drc_state *state = &cd->state;
+	int ch, err;
+
+	if (!config) {
+		comp_cl_err(&comp_multiband_drc, "crossover_eq_init_coef(), no config is set");
+		return -EINVAL;
+	}
+
+	/* Sanity checks */
+	if (nch > PLATFORM_MAX_CHANNELS) {
+		comp_cl_err(&comp_multiband_drc,
+			    "crossover_eq_init_coef(), invalid channels count(%i)",
+			    nch);
+		return -EINVAL;
+	}
+
+	comp_cl_info(&comp_multiband_drc, "crossover_eq_init_coef(), initiliazing %i-way crossover",
+		     config->num_bands);
+
+	/* Crossover: collect the coef array and assign it to every channel */
+	crossover = config->crossover_coef;
+	for (ch = 0; ch < nch; ch++) {
+		err = crossover_init_coef_ch(crossover, &state->crossover[ch],
+					     config->num_bands);
+		/* Free all previously allocated blocks in case of an error */
+		if (err < 0) {
+			comp_cl_err(&comp_multiband_drc,
+				    "crossover_eq_init_coef(), could not assign coeffs to ch %d",
+				    ch);
+			multiband_drc_crossover_reset_state(state);
+			return err;
+		}
+	}
+
+	comp_cl_info(&comp_multiband_drc, "crossover_eq_init_coef(), initiliazing emphasis_eq");
+
+	/* Emphasis: collect the coef array and assign it to every channel */
+	emphasis = config->emp_coef;
+	for (ch = 0; ch < nch; ch++) {
+		err = multiband_drc_eq_init_coef_ch(emphasis, &state->emphasis[ch]);
+		/* Free all previously allocated blocks in case of an error */
+		if (err < 0) {
+			comp_cl_err(&comp_multiband_drc,
+				    "crossover_eq_init_coef(), could not assign coeffs to ch %d",
+				    ch);
+			multiband_drc_emphasis_reset_state(state);
+			return err;
+		}
+	}
+
+	comp_cl_info(&comp_multiband_drc, "crossover_eq_init_coef(), initiliazing deemphasis_eq");
+
+	/* Deemphasis: collect the coef array and assign it to every channel */
+	deemphasis = config->deemp_coef;
+	for (ch = 0; ch < nch; ch++) {
+		err = multiband_drc_eq_init_coef_ch(deemphasis, &state->deemphasis[ch]);
+		/* Free all previously allocated blocks in case of an error */
+		if (err < 0) {
+			comp_cl_err(&comp_multiband_drc,
+				    "crossover_eq_init_coef(), could not assign coeffs to ch %d",
+				    ch);
+			multiband_drc_deemphasis_reset_state(state);
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static int multiband_drc_drc_init(struct multiband_drc_comp_data *cd, int16_t nch, uint32_t rate)
+{
+	uint32_t sample_bytes = get_sample_bytes(cd->source_format);
+	int num_bands = cd->config->num_bands;
+	struct multiband_drc_state *state = &cd->state;
+
+	int i;
+	int ret;
+
+	/* Allocate all DRC pre-delay buffers and set delay time with band number */
+	for (i = 0; i < SOF_MULTIBAND_DRC_MAX_BANDS; i++) {
+		if (i == num_bands)
```

Comment:
I don't see a check that num_bands is less than SOF_MULTIBAND_DRC_MAX_BANDS. If such has been done earlier, the above loop could be with just num_bands.